pinot-plugins/pinot-segment-writer/pinot-segment-writer-file-based/src/main/java/org/apache/pinot/plugin/segmentwriter/filebased/FileBasedSegmentWriter.java

```diff
@@ -0,0 +1,197 @@
+    _bufferFile = new File(bufferDir, "buffer_file");
+    resetBuffer();
+    LOGGER.info("Initialized {} for table: {}", FileBasedSegmentWriter.class.getName(), _tableNameWithType);
+  }
+
+  private void resetBuffer()
+      throws IOException {
+    FileUtils.deleteQuietly(_bufferFile);
+    _recordWriter = new DataFileWriter<>(new GenericDatumWriter<>(_avroSchema));
+    _recordWriter.create(_avroSchema, _bufferFile);
+  }
+
+  @Override
+  public void collect(GenericRow row)
+      throws IOException {
+    GenericRow transform = _recordTransformer.transform(row);
+    SegmentProcessorAvroUtils.convertGenericRowToAvroRecord(transform, _reusableRecord, _fieldsToRead);
+    _recordWriter.append(_reusableRecord);
+  }
+
+  @Override
+  public URI flush()
+      throws IOException {
+
+    LOGGER.info("Beginning flush for table: {}", _tableNameWithType);
+    _recordWriter.close();
+
+    // Create temp dir for flush
+    File flushDir = new File(_stagingDir, "flush_dir_" + System.currentTimeMillis());
+    Preconditions.checkState(flushDir.mkdirs(), "Failed to create flush dir: %s", flushDir);
+
+    try {
+      // Segment dir
+      File segmentDir = new File(flushDir, "segment_dir");
+
+      // Make BatchIngestionConfig for flush
+      Map<String, String> batchConfigMapOverride = new HashMap<>(_batchConfig.getBatchConfigMap());
+      batchConfigMapOverride.put(BatchConfigProperties.INPUT_DIR_URI, _bufferFile.getAbsolutePath());
+      batchConfigMapOverride.put(BatchConfigProperties.OUTPUT_DIR_URI, segmentDir.getAbsolutePath());
+      batchConfigMapOverride.put(BatchConfigProperties.INPUT_FORMAT, BUFFER_FILE_FORMAT.toString());
+      BatchIngestionConfig batchIngestionConfig = new BatchIngestionConfig(Lists.newArrayList(batchConfigMapOverride),
+          _batchIngestionConfig.getSegmentIngestionType(), _batchIngestionConfig.getSegmentIngestionFrequency());
+
+      // Build segment
+      SegmentGeneratorConfig segmentGeneratorConfig =
+          IngestionUtils.generateSegmentGeneratorConfig(_tableConfig, _schema, batchIngestionConfig);
+      String segmentName = IngestionUtils.buildSegment(segmentGeneratorConfig);
+      LOGGER.info("Successfully built segment: {} for table: {}", segmentName, _tableNameWithType);
+
+      // Tar segment
+      File segmentTarFile = new File(_outputDirURI, segmentName + Constants.TAR_GZ_FILE_EXT);
+      if (!_batchConfig.isOverwriteOutput() && segmentTarFile.exists()) {
+        segmentTarFile = new File(_outputDirURI,
+            String.format("%s_%d%s", segmentName, System.currentTimeMillis(), Constants.TAR_GZ_FILE_EXT));
+      }
+      TarGzCompressionUtils.createTarGzFile(new File(segmentDir, segmentName), segmentTarFile);
+      LOGGER.info("Created segment tar: {} for segment: {} of table: {}", segmentTarFile.getAbsolutePath(), segmentName,
+          _tableNameWithType);
+
+      // Reset buffer and return segmentTar URI
+      resetBuffer();
```

Comment:
Should `resetBuffer` be in finally ? Given that we've already close _recordWriter ?

Reply:
The reason I didn't put it in finally, is that if the `flush` returns an exception to the caller, the caller should be able to invoke `flush` again. They'd probably have `flush` in a retry loop.  If reset goes into finally, we will lose the buffered data.
But I see where you're coming from. If caller gets an error from `flush`, they would want to start rewriting the records? I was thinking that they should init the SegmentWriter again in such a case.

Reply:
I think we should probably make the semantics clear in either case (maybe via interface java docs). I think its fine to assume caller will retry.

+yupeng9 for visibility

Reply:
Added the javadocs to make the contract clear

Reply:
Thanks for calling this out.
Is it possible to tell which exceptions are retriable and which are not? Or have an argument of `retryNum` so the retries are done within the implementation?